File: test/node/helper/apiHelpers.ts
/**
 * @fileoverview Manual API Request Utilities for DANA Integration Testing
 * 
 * This module provides utilities for making direct API requests with custom headers
 * and authentication mechanisms. It's primarily used for testing edge cases where
 * the standard SDK behavior needs to be bypassed, such as testing invalid signatures,
 * missing headers, or custom authentication scenarios.
 * 
 * Key Features:
 * - Direct HTTP API calls bypassing SDK validation
 * - Custom header injection capabilities
 * - Manual signature generation and validation
 * - Support for negative testing scenarios
 */

import { v4 as uuidv4 } from 'uuid';
import dotenv from 'dotenv';
import axios from 'axios';
import { BaseAPI, Configuration, DanaSignatureUtil } from 'dana-node/runtime';
import { HTTPMethod } from 'dana-node/runtime';

// Load environment variables from .env file
dotenv.config();

/**
 * Executes an API request directly with manually constructed headers and request parameters
 * 
 * This function provides direct access to the DANA API by constructing HTTP requests
 * with custom headers and authentication. It's designed for testing scenarios where
 * the standard SDK behavior needs to be bypassed or modified, such as:
 * - Testing with invalid or missing authentication headers
 * - Validating API behavior with malformed requests
 * - Testing edge cases and error handling scenarios
 * 
 * The function handles signature generation using the DANA signature utility and
 * allows for custom header injection to override default values.
 * 
 * @param {string} caseName - Test case name for logging and debugging purposes
 * @param {string} method - HTTP method (e.g., "POST", "GET", "PUT", "DELETE")
 * @param {string} endpoint - Complete API endpoint URL including domain
 * @param {string} resourcePath - API resource path used for signature generation (e.g., "/payment-gateway/v1.0/debit/create.htm")
 * @param {Record<string, any>} requestObj - Request payload object to send in the request body
 * @param {Record<string, string>} customHeaders - Custom headers to include or override defaults (default: {})
 * @returns {Promise<ManualApiResponse>} Response object containing data, status, headers, and raw response
 * @throws {Error} Throws error if signature generation fails or request execution fails
 * 
 * @example
 * ```typescript
 * // Test with invalid signature
 * const response = await executeManualApiRequest(
 *   "TestInvalidSignature",
 *   "POST",
 *   "https://api.sandbox.dana.id/payment-gateway/v1.0/debit/create.htm",
 *   "/payment-gateway/v1.0/debit/create.htm",
 *   requestData,
 *   { 'X-SIGNATURE': 'invalid_signature_for_testing' }
 * );
 * 
 * // Test with missing timestamp
 * const response = await executeManualApiRequest(
 *   "TestMissingTimestamp",
 *   "POST",
 *   "https://api.sandbox.dana.id/payment-gateway/v1.0/debit/cancel.htm",
 *   "/payment-gateway/v1.0/debit/cancel.htm",
 *   requestData,
 *   { 'X-TIMESTAMP': '' }
 * );
 * ```
 */
async function executeManualApiRequest(
  caseName: string,
  method: string,
  endpoint: string,
  resourcePath: string,
  requestObj: Record<string, any>,
  customHeaders: Record<string, string> = {}
): Promise<{ data: any; status: number; headers: any; rawResponse: any }> {

  // Initialize default headers required for DANA API authentication
  const headers: Record<string, string> = {
    'X-PARTNER-ID': process.env.X_PARTNER_ID || '',
    'CHANNEL-ID': process.env.CHANNEL_ID || '',
    'Content-Type': 'application/json',
    'ORIGIN': process.env.ORIGIN || '',
    'X-EXTERNAL-ID': uuidv4() // Generate unique external ID for request tracking
  };

  // Generate timestamp in Jakarta timezone (UTC+7) format
  const now = new Date();
  const offset = '+07:00'; // Jakarta timezone offset
  const isoString = now.toISOString();
  const timeStamp = isoString.substring(0, 19).replace('Z', '') + offset;

  headers['X-TIMESTAMP'] = timeStamp;

  // Generate DANA signature for authentication using the SDK utility
  const requestBody = JSON.stringify(requestObj);
  const snapGeneratedHeaders = DanaSignatureUtil.generateSnapB2BScenarioSignature(
    method,
    resourcePath,
    requestBody,
    process.env.PRIVATE_KEY || '',
    timeStamp
  );

  if (snapGeneratedHeaders) {
    headers['X-SIGNATURE'] = snapGeneratedHeaders;
  } else {
    throw new Error('Signature generation failed - check private key and parameters');
  }

  // Apply custom headers (these will override default values for testing purposes)
  if (customHeaders) {
    Object.keys(customHeaders).forEach(key => {
      headers[key] = customHeaders[key];
    });
  }

  // Log request details for debugging
  console.log(`Executing manual API request for case: ${caseName}`);
  console.log(`Method: ${method}, Endpoint: ${endpoint}`);
  console.log(`Headers:`, headers);

  // Alternative HTTP implementation using axios (commented out but available for reference)
  // const response = await axios({
  //   method: method,
  //   url: endpoint,
  //   headers: headers,
  //   data: requestObj,
  //   validateStatus: function (status) {
  //     // Return true for any status code (don't throw errors)
  //     return true;
  //   }
  // });

  // Create DANA SDK configuration for API client
  const configParams = {
    basePath: 'https://api.sandbox.dana.id'
  };

  const config = new Configuration(configParams);
  const api_client = new BaseAPI(config);

  // Execute the API request using the DANA SDK's underlying HTTP client
  // @ts-ignore - Ignore private property access warnings for testing purposes
  const response = await api_client.request({
    method: method as HTTPMethod,
    path: resourcePath,
    headers: headers,
    body: requestObj
  });

  // Return standardized response object for test validation
  return {
    data: response.body,
    status: response.status,
    headers: response.headers,
    rawResponse: response.body // Include raw response for detailed error analysis
  };
}

/**
 * Export manual API request utilities
 * 
 * These exports provide direct API access capabilities for advanced testing scenarios:
 * - Manual request execution with custom headers
 * - Signature generation bypassing
 * - Error scenario simulation
 */
export {
  executeManualApiRequest,
};

File: test/python/helper/api_helpers.py
import json
import os
from uuid import uuid4
from datetime import datetime, timezone, timedelta
import pytest

from helper.assertion import assert_fail_response

def execute_api_request_directly(api_client, method, endpoint, request_obj, headers):
    """
    Executes a direct API call with the provided headers and request object.
    This is a direct wrapper around api_client.call_api that preserves the exact behavior.
    
    :param api_client: The API client to use for the request
    :param method: HTTP method (e.g., "POST", "GET")
    :param endpoint: The API endpoint URL
    :param request_obj: The request object to send
    :param headers: The exact headers to use
    :return: The API response
    """
    # Convert request object to dictionary if needed
    request_data = request_obj
    if hasattr(request_obj, "to_dict"):
        request_data = request_obj.to_dict()
    
    # HTTP requires header values to be strings; env-based headers (e.g. CLIENT_SECRET) may be None
    headers = {k: (v if v is not None else "") for k, v in headers.items()}
    
    # Execute the API call with exactly the provided headers
    return api_client.call_api(method, endpoint, headers, request_data)


def get_standard_headers(with_timestamp=True):
    """
    Returns standard headers used in API requests.
    
    :param with_timestamp: Whether to include X-TIMESTAMP (defaults to True)
    :return: Dict of standard headers
    """
    headers = {
        "X-PARTNER-ID": os.getenv("X_PARTNER_ID"),
        "CHANNEL-ID": "95221",
        "ORIGIN": os.getenv("ORIGIN"),
        "X-EXTERNAL-ID": str(uuid4()),
        "CLIENT_SECRET": os.getenv("CLIENT_SECRET"),
        "Content-Type": "application/json"
    }
    
    if with_timestamp:
        headers["X-TIMESTAMP"] = datetime.now().astimezone(
            timezone(timedelta(hours=7))
        ).strftime('%Y-%m-%dT%H:%M:%S+07:00')
    
    return headers


def get_headers_with_signature(method=None, resource_path=None, request_obj=None, with_timestamp=True, invalid_timestamp=False, invalid_signature=False):
    """
    Creates headers including the signature, handling all the signature generation internally
    
    :param method: HTTP method (e.g., "POST", "GET") - required unless invalid_signature=True
    :param resource_path: API resource path (e.g., "/payment-gateway/v1.0/debit/status.htm") - required unless invalid_signature=True
    :param request_obj: Request object or dictionary - required unless invalid_signature=True
    :param with_timestamp: Whether to include X-TIMESTAMP (defaults to True)
    :param invalid_timestamp: If True, uses an invalid timestamp format
    :param invalid_signature: If True, uses an invalid signature
    :return: Dict of headers including the signature
    """
    from dana.utils.snap_header import SnapHeader
    import json
    import os
    
    # Get the standard headers first
    headers = get_standard_headers(with_timestamp=False)  # We'll handle timestamp specially
    
    # Add the signature
    if invalid_signature:
        headers["X-SIGNATURE"] = "85be817c55b2c135157c7e89f52499bf0c25ad6eeebe04a986e8c862561b19a5"  # Invalid signature
    elif invalid_signature is None:
        headers["X-SIGNATURE"] = ""
        print("Warning: Invalid signature, using None")
    else:
        if method is None or resource_path is None or request_obj is None:
            raise ValueError("Method, resource_path, and request_obj are required unless invalid_signature=True")
            
        # Convert request object to dict if needed
        request_dict = request_obj
        if hasattr(request_obj, "to_dict"):
            request_dict = request_obj.to_dict()
            
        # Generate snap headers
        snap_generated_headers = SnapHeader.get_snap_generated_auth(
            method=method,
            resource_path=resource_path,
            body=json.dumps(request_dict),
            private_key=os.getenv("PRIVATE_KEY")
        )
        
        headers["X-SIGNATURE"] = snap_generated_headers["X-SIGNATURE"]["value"]
    
    # Add timestamp if needed
    if with_timestamp:
        if invalid_timestamp:
            headers["X-TIMESTAMP"] = datetime.now().astimezone(
                timezone(timedelta(hours=7))
            ).strftime('%Y-%m-%d %H:%M:%S+07:00')  # Invalid format (space instead of T)
        else:
            headers["X-TIMESTAMP"] = datetime.now().astimezone(
                timezone(timedelta(hours=7))
            ).strftime('%Y-%m-%dT%H:%M:%S+07:00')  # Valid format
    
    return headers


def execute_and_assert_api_error(api_client, method, endpoint, request_obj, 
                                headers, expected_status, 
                                json_path_file, title_case, case_name, 
                                variable_dict=None):
    """
    Executes an API request and asserts the error response matches expectations.
    This maintains the exact same implementation that was originally in the test files.
    
    :param api_client: The API client instance
    :param method: HTTP method (POST, GET, etc)
    :param endpoint: API endpoint URL
    :param request_obj: Request object or dictionary 
    :param headers: Complete headers dictionary to use
    :param expected_status: Expected HTTP status code
    :param json_path_file: Path to JSON file with expected responses
    :param title_case: Title case for the test (e.g. "QueryPayment")
    :param case_name: Specific test case name
    :param variable_dict: Optional dictionary for variable substitution
    """
    try:
        # Make the API call with the exact provided headers
        response = execute_api_request_directly(
            api_client, 
            method, 
            endpoint, 
            request_obj, 
            headers
        )

        # Assert the response status matches expectations
        try:
            assert response.status == expected_status
        except:
            pytest.fail(f"Expected status {expected_status} but got {response.status}")
        
        # Read the response data and decode from bytes to string
        response_data = response.read().decode('utf-8')
        
        # Assert the API response
        assert_fail_response(json_path_file, title_case, case_name, response_data, variable_dict)

    except Exception as e:
        # If the exception has a desired attribute, use it directly
        if hasattr(e, "body") and e.body:
            assert_fail_response(json_path_file, title_case, case_name, e.body, variable_dict)
        else:
            pytest.fail(f"Failed to call API: {str(e)}")
